File: frontend/src/api/billing.ts
import axiosInstance from './axios-instance';

export type BillingItemStatus = 'pending' | 'paid' | 'cancelled';
export type BillingItemKind = 'hosting' | 'domain' | 'other';
export type BillingPeriod = 'month' | 'year' | 'one_time';
export type BillingCheckoutMode = 'subscription' | 'prepay_year';

export interface BillingItem {
    id: string;
    user_id: string;
    resource_type: BillingItemKind;
    resource_id: string | null;
    title: string;
    description: string | null;
    amount_cents: number;
    currency: string;
    billing_period: BillingPeriod;
    status: BillingItemStatus;
    due_at: string;
    grace_period_ends_at: string;
    paid_at: string | null;
    stripe_session_id: string | null;
    metadata: Record<string, unknown> | null;
    created_at: string;
    updated_at: string;
}

export interface CreateBillingCheckoutRequest {
    item_ids?: string[];
    mode: BillingCheckoutMode;
}

export interface BillingCheckoutResponse {
    checkout_url: string;
}

export async function apiListBillingItems(): Promise<BillingItem[]> {
    const {data} = await axiosInstance.get<BillingItem[]>('/api/billing/items');
    return data;
}

export async function apiCreateBillingCheckout(
    req: CreateBillingCheckoutRequest,
): Promise<string> {
    const {data} = await axiosInstance.post<BillingCheckoutResponse>('/api/billing/checkout', req);
    return data.checkout_url;
}

export function billingItemPeriodLabel(period: BillingPeriod): string {
    switch (period) {
    case 'month':
        return '/mes';
    case 'year':
        return '/año';
    default:
        return '';
    }
}

File: src/models/billing.rs
/* [205A-1] Facturacion pendiente desacoplada de hosting_subscriptions.
 * Un hosting puede seguir activo mientras la plataforma muestra el cobro que falta activar. */
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow, Serialize, ToSchema)]
pub struct BillingItem {
    pub id: Uuid,
    pub user_id: Uuid,
    pub resource_type: String,
    pub resource_id: Option<Uuid>,
    pub title: String,
    pub description: Option<String>,
    pub amount_cents: i32,
    pub currency: String,
    pub billing_period: String,
    pub status: String,
    pub due_at: DateTime<Utc>,
    pub grace_period_ends_at: DateTime<Utc>,
    pub paid_at: Option<DateTime<Utc>>,
    pub stripe_session_id: Option<String>,
    pub metadata: serde_json::Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum BillingCheckoutMode {
    Subscription,
    PrepayYear,
}

impl std::fmt::Display for BillingCheckoutMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Subscription => write!(f, "subscription"),
            Self::PrepayYear => write!(f, "prepay_year"),
        }
    }
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct CreateBillingCheckoutRequest {
    pub item_ids: Option<Vec<Uuid>>,
    pub mode: BillingCheckoutMode,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct BillingCheckoutResponse {
    pub checkout_url: String,
}

/* [026B-1] Respuesta admin de billing_items con email del usuario. */
#[derive(Debug, Serialize, ToSchema)]
pub struct AdminBillingItemResponse {
    pub id: Uuid,
    pub user_id: Uuid,
    pub user_email: String,
    pub resource_type: String,
    pub resource_id: Option<Uuid>,
    pub title: String,
    pub description: Option<String>,
    pub amount_cents: i32,
    pub currency: String,
    pub billing_period: String,
    pub status: String,
    pub due_at: DateTime<Utc>,
    pub grace_period_ends_at: DateTime<Utc>,
    pub paid_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/* [026B-1] Request para cambiar status de un billing_item desde admin. */
#[derive(Debug, Deserialize, ToSchema)]
pub struct AdminUpdateBillingStatusRequest {
    pub status: String,
}
